File: apps/web/core/types/workspace-feature.ts
/**
 * SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
 * SPDX-License-Identifier: LicenseRef-Plane-Commercial
 *
 * Licensed under the Plane Commercial License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://plane.so/legals/eula
 *
 * DO NOT remove or modify this notice.
 * NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.
 */

export enum EWorkspaceFeatureLoader {
  INIT_LOADER = "workspace-feature-init-loader",
  MUTATION_LOADER = "workspace-feature-mutation-loader",
}

export type TWorkspaceFeatureLoader = EWorkspaceFeatureLoader | undefined;

// workspace feature
export enum EWorkspaceFeatures {
  IS_PROJECT_GROUPING_ENABLED = "is_project_grouping_enabled",
  IS_TEAMSPACES_ENABLED = "is_teams_enabled",
  IS_INITIATIVES_ENABLED = "is_initiative_enabled",
  IS_CUSTOMERS_ENABLED = "is_customer_enabled",
  IS_PI_ENABLED = "is_pi_enabled",
  IS_WIKI_ENABLED = "is_wiki_enabled",
  IS_MILESTONES_ENABLED = "is_milestones_enabled",
}

export type TWorkspaceFeature = { [key in EWorkspaceFeatures]: boolean | undefined };

export type TWorkspaceFeatures = {
  workspace: string | undefined;
  created_at: string | undefined;
  updated_at: string | undefined;
} & TWorkspaceFeature;

File: apps/api/plane/ee/utils/workspace_feature.py
# SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
# SPDX-License-Identifier: LicenseRef-Plane-Commercial
#
# Licensed under the Plane Commercial License (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# https://plane.so/legals/eula
#
# DO NOT remove or modify this notice.
# NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.

# Third party imports
from enum import Enum

from plane.ee.models.workspace import WorkspaceFeature


class WorkspaceFeatureContext(Enum):
    # Workspace level project states
    IS_PROJECT_GROUPING_ENABLED = "is_project_grouping_enabled"
    IS_CUSTOMER_ENABLED = "is_customer_enabled"
    IS_WORK_ITEM_TYPES_ENABLED = "is_work_item_types_enabled"


def check_workspace_feature(slug, feature: WorkspaceFeatureContext):
    # Dynamically build the filter using the feature's value
    filter_kwargs = {"workspace__slug": slug, feature.value: True}
    is_workspace_feature_enabled = WorkspaceFeature.objects.filter(**filter_kwargs).exists()

    return is_workspace_feature_enabled
